File: bcw5/php/PenPaper/Paper.php
<?php
	class Paper {
        private $maxSymbols;
        private $symbols;
        private $content;

        public function Paper($maxSymbols=4096) {
        	$this->maxSymbols = $maxSymbols;
		    $this->symbols = 0;
		    $this->content = "";
        }

        public function getMaxSymbols() {
    		return $this->maxSymbols;
        }

        public function getSymbols() {
        	return $this->symbols;
        }

        public function addContent($message) {
		    $total = $this->symbols + strlen($message);

		    if ( $this->symbols == $this->maxSymbols ) {
		        throw new Exception("OutOfSpace");
		    }
		    if ( $total > $this->maxSymbols ) {
		        $len = $this->maxSymbols - $this->symbols;

		        $this->content += substr($message, $len - 1);
		        $this->symbols = $this->maxSymbols;
		        return;
		    }
		    $this->content = $this->content . $message;
		    $this->symbols += strlen($message);
        }

        public function show() {
        	echo "Message: " . $this->content . PHP_EOL;
        }

        public function output() {
        	echo PHP_EOL . "My Piece of paper properties: " . PHP_EOL .
        	"Symbols on it: " . $this->symbols . " of " . $this->maxSymbols . PHP_EOL;
        }
	}
?>

File: bcw4/java/PenPaper/Paper.java
public class Paper {
    private int maxSymbols;
    private int symbols;
    private StringBuffer content;

    public Paper() {
        this.maxSymbols = 4096;
        this.symbols = 0;
        content = new StringBuffer();
    }

    public Paper(int maxSymbols) {
        this.maxSymbols = maxSymbols;
        this.symbols = 0;
        this.content = new StringBuffer();
    }

    public int getMaxSymbols() {
        return maxSymbols;
    }

    public int getSymbols() {
        return symbols;
    }

    public void addContent(String message) throws OutOfSpaceException {
        int total = content.length() + message.length();

        if ( content.length() == maxSymbols ) {
            throw new OutOfSpaceException();
        }
        if ( total > maxSymbols ) {
            content.append(message.substring(0, maxSymbols-content.length()));
            return;
        }
        content.append(message);
        symbols += content.length();
    }

    public void show() {
        System.out.println(content);
    }
};